File: tools/pw-grants-crawler/src/pw_grants_crawler/renderer.py
from .models import PageSnapshot


class PlaywrightFetcher:
    """Render pages that cannot be understood from the static HTTP response."""

    def __init__(
        self,
        *,
        timeout: float = 30.0,
        max_bytes: int = 5_000_000,
        user_agent: str = "Missa-PW-Grants-Crawler/0.1 (+https://usemissa.com)",
    ):
        try:
            from playwright.sync_api import sync_playwright
        except ImportError as exc:
            raise RuntimeError(
                "Rendering requires the optional dependency. Run: "
                "uv sync --extra render && uv run playwright install chromium"
            ) from exc

        self.timeout_ms = int(timeout * 1000)
        self.max_bytes = max_bytes
        self._playwright = sync_playwright().start()
        try:
            self._browser = self._playwright.chromium.launch(headless=True)
            self._context = self._browser.new_context(user_agent=user_agent)
        except Exception:
            self._playwright.stop()
            raise

    def fetch(self, url: str) -> PageSnapshot:
        page = self._context.new_page()
        resource_urls: set[str] = set()
        resource_payloads: dict[str, tuple[str | None, bytes]] = {}

        def record_resource(response) -> None:
            if response.request.resource_type != "image":
                return
            resource_urls.add(response.url)
            try:
                body = response.body()
            except Exception:
                return
            if len(body) <= self.max_bytes:
                resource_payloads[response.url] = (
                    response.headers.get("content-type"),
                    body,
                )

        page.on("response", record_resource)
        try:
            response = page.goto(url, wait_until="domcontentloaded", timeout=self.timeout_ms)
            try:
                page.wait_for_load_state("networkidle", timeout=min(self.timeout_ms, 10_000))
            except Exception:
                pass
            html = page.content()
            if len(html.encode("utf-8")) > self.max_bytes:
                return PageSnapshot(
                    requested_url=url,
                    final_url=page.url,
                    status_code=response.status if response else None,
                    content_type="text/html",
                    html="",
                    title="",
                    text="",
                    error=f"Rendered response exceeded max_bytes={self.max_bytes}",
                    rendered=True,
                    resource_urls=sorted(resource_urls),
                    resource_payloads=resource_payloads,
                )
            return PageSnapshot(
                requested_url=url,
                final_url=page.url,
                status_code=response.status if response else None,
                content_type="text/html",
                html=html,
                title=page.title(),
                text=page.locator("body").inner_text(timeout=self.timeout_ms),
                error=None if response is None or response.ok else f"HTTP {response.status}",
                rendered=True,
                resource_urls=sorted(resource_urls),
                resource_payloads=resource_payloads,
            )
        except Exception as exc:
            return PageSnapshot(
                requested_url=url,
                final_url=page.url or url,
                status_code=None,
                content_type=None,
                html="",
                title="",
                text="",
                error=str(exc),
                rendered=True,
                resource_urls=sorted(resource_urls),
                resource_payloads=resource_payloads,
            )
        finally:
            page.close()

    def close(self) -> None:
        self._context.close()
        self._browser.close()
        self._playwright.stop()

File: packages/render-service/src/renderer.ts
import { chromium, type Browser } from "playwright";
import { assessRenderUrl } from "./policy.js";

export interface RenderResult {
  url: string;
  finalUrl: string;
  statusCode: number;
  contentType: string | null;
  html: string;
  rendered: true;
}

export interface RendererOptions {
  userAgent?: string;
  timeoutMs?: number;
  maxBytes?: number;
}

const DEFAULT_USER_AGENT = "MissaIngestionV2/0.1 (+https://www.usemissa.com; rendering)";

/**
 * One shared browser, one fresh context per request. A context per request keeps
 * cookies and storage from leaking between organizations; a shared browser keeps
 * us from paying Chromium's start-up cost on every page.
 */
export class Renderer {
  private browser?: Browser;
  private readonly userAgent: string;
  private readonly timeoutMs: number;
  private readonly maxBytes: number;

  constructor(options: RendererOptions = {}) {
    this.userAgent = options.userAgent ?? DEFAULT_USER_AGENT;
    this.timeoutMs = options.timeoutMs ?? 20_000;
    this.maxBytes = options.maxBytes ?? 5_000_000;
  }

  async render(target: string): Promise<RenderResult> {
    const decision = assessRenderUrl(target);
    if (!decision.allowed) throw Object.assign(new Error(decision.reason), { statusCode: 400 });

    const browser = await this.browser0();
    const context = await browser.newContext({ userAgent: this.userAgent, javaScriptEnabled: true });
    const page = await context.newPage();
    // Images and fonts cost time and bytes and never carry the text we extract.
    await page.route("**/*", (route) => {
      const type = route.request().resourceType();
      return type === "image" || type === "media" || type === "font" ? route.abort() : route.continue();
    });
    try {
      const response = await page.goto(decision.url.href, { waitUntil: "domcontentloaded", timeout: this.timeoutMs });
      if (!response) throw Object.assign(new Error("The page returned no response"), { statusCode: 502 });
      await page.waitForLoadState("networkidle", { timeout: Math.min(this.timeoutMs, 3_000) }).catch(() => undefined);
      const html = await page.content();
      if (html.length > this.maxBytes) throw Object.assign(new Error("The rendered document exceeded the size limit"), { statusCode: 413 });
      return {
        url: decision.url.href,
        finalUrl: page.url(),
        statusCode: response.status(),
        contentType: response.headers()["content-type"] ?? null,
        html,
        rendered: true,
      };
    } finally {
      await context.close().catch(() => undefined);
    }
  }

  async close(): Promise<void> {
    await this.browser?.close().catch(() => undefined);
    this.browser = undefined;
  }

  private async browser0(): Promise<Browser> {
    if (!this.browser || !this.browser.isConnected()) {
      this.browser = await chromium.launch({ headless: true, args: ["--no-sandbox", "--disable-dev-shm-usage"] });
    }
    return this.browser;
  }
}
